pkg/clustermanager/provider.go

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2023 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clustermanager
+
+import (
+	"context"
+	"errors"
+	"strings"
+
+	"github.com/go-logr/logr"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/klog/v2"
+
+	kindprovider "github.com/kubestellar/kubestellar/clusterprovider/kind"
+	lcv1alpha1apis "github.com/kubestellar/kubestellar/pkg/apis/logicalcluster/v1alpha1"
+	clusterprovider "github.com/kubestellar/kubestellar/pkg/clustermanager/providerclient"
+)
+
+// Each provider gets its own namespace named prefixNamespace+providerName
+const prefixNamespace = "lcprovider-"
+
+func ProviderNS(name string) string {
+	return prefixNamespace + name
+}
+
+func lcKeyFunc(ns string, name string) string {
+	if ns != "" {
+		return ns + "/" + name
+	}
+	return name
+}
+
+type provider struct {
+	name            string
+	providerClient  clusterprovider.ProviderClient
+	c               *controller
+	providerWatcher clusterprovider.Watcher
+	nameSpace       string
+	discoveryPrefix string
+}
+
+// TODO: this is termporary for stage 1. For stage 2 we expect to have a uniform interface for all informers.
+func newProviderClient(providerName string, providerType lcv1alpha1apis.ClusterProviderType) clusterprovider.ProviderClient {
+	var pClient clusterprovider.ProviderClient = nil
+	switch providerType {
+	case lcv1alpha1apis.KindProviderType:
+		pClient = kindprovider.New(providerName)
+	default:
+		return nil
+	}
+	return pClient
+}
+
+// CreateProvider returns new provider client
+func CreateProvider(c *controller, providerName string, providerType lcv1alpha1apis.ClusterProviderType) (*provider, error) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	discoveryPrefix := "*"
```

Comment:
Wouldn't it be simpler to use the empty string as the default?

Reply:
Handled by PR #816